client/funcopts: Migrate `ImageHistory`, `ImageInspect` and `ContainerStats`

vvoland opened the pull request:
- alternative to: https://github.com/moby/moby/pull/50932

This moves the options to separate packages (per each operation).

This makes the Godoc a bit cleaner, but doesn't make it possible to use shared options (we could, but we'd need to put them in a separate package which wouldn't make them visible in the operation-specific package).


<img width="1599" height="777" alt="image" src="https://github.com/user-attachments/assets/ff843098-31bb-4ea9-bb1b-c98291dfc0d7" />


<img width="1567" height="959" alt="image" src="https://github.com/user-attachments/assets/db604c85-b062-4c63-bef8-972a15b99bc1" />

vvoland committed:
client/funcopts: Migrate `ImageHistory` and `ImageInspect`

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented:
What if the type definition was also in the `imageopt` package?

```go
client.ImageInspect(ctx, imageopt.InspectOption)
```

Then clicking `imageopt.InspectOption` would bring you to the `imageopt` package, with all options listed;

<img width="1431" height="1149" alt="Screenshot 2025-09-19 at 18 31 37" src="https://github.com/user-attachments/assets/4fe90251-cb22-485c-9f8e-5499fd0d79c1" />

That `imageopt` package would also include the type as navigation on the left-hand side, which can be expanded to show all the types related to it;

<img width="1431" height="1149" alt="Screenshot 2025-09-19 at 18 32 02" src="https://github.com/user-attachments/assets/4227cd0a-60ff-40e9-90d2-70024e049573" />

thaJeztah commented:
This is the patch I tried;

```patch
From f3fa99da5a76e78b1854089fb32434e09d7f4294 Mon Sep 17 00:00:00 2001
From: Sebastiaan van Stijn <github@gone.nl>
Date: Fri, 19 Sep 2025 18:37:01 +0200
Subject: [PATCH] move the option type

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
---
 client/client_interfaces.go  |  3 ++-
 client/image_inspect.go      |  8 ++------
 client/image_inspect_test.go |  3 ++-
 client/imageopt/inspect.go   |  5 ++---
 client/imageopt/opt.go       | 12 ++++++++++++
 client/imageopt/platform.go  |  3 +--
 6 files changed, 21 insertions(+), 13 deletions(-)
 create mode 100644 client/imageopt/opt.go

diff --git a/client/client_interfaces.go b/client/client_interfaces.go
index ae6e529cc8..8374e923ff 100644
--- a/client/client_interfaces.go
+++ b/client/client_interfaces.go
@@ -18,6 +18,7 @@ import (
 	"github.com/moby/moby/api/types/swarm"
 	"github.com/moby/moby/api/types/system"
 	"github.com/moby/moby/api/types/volume"
+	"github.com/moby/moby/client/imageopt"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
@@ -124,7 +125,7 @@ type ImageAPIClient interface {
 	ImageTag(ctx context.Context, image, ref string) error
 	ImagesPrune(ctx context.Context, pruneFilter filters.Args) (image.PruneReport, error)
 
-	ImageInspect(ctx context.Context, image string, _ ...ImageInspectOption) (image.InspectResponse, error)
+	ImageInspect(ctx context.Context, image string, _ ...imageopt.InspectOption) (image.InspectResponse, error)
 	ImageHistory(ctx context.Context, image string, _ ...ImageHistoryOption) ([]image.HistoryResponseItem, error)
 	ImageLoad(ctx context.Context, input io.Reader, _ ...ImageLoadOption) (LoadResponse, error)
 	ImageSave(ctx context.Context, images []string, _ ...ImageSaveOption) (io.ReadCloser, error)
diff --git a/client/image_inspect.go b/client/image_inspect.go
index 8236cb13a7..8043c078b5 100644
--- a/client/image_inspect.go
+++ b/client/image_inspect.go
@@ -9,16 +9,12 @@ import (
 	"net/url"
 
 	"github.com/moby/moby/api/types/image"
+	"github.com/moby/moby/client/imageopt"
 	"github.com/moby/moby/client/internal/opts"
 )
 
-// ImageInspectOption is a type representing functional options for the image inspect operation.
-type ImageInspectOption interface {
-	ApplyImageInspectOption(ctx context.Context, opts *opts.ImageInspectOptions) error
-}
-
 // ImageInspect returns the image information.
-func (cli *Client) ImageInspect(ctx context.Context, imageID string, inspectOpts ...ImageInspectOption) (image.InspectResponse, error) {
+func (cli *Client) ImageInspect(ctx context.Context, imageID string, inspectOpts ...imageopt.InspectOption) (image.InspectResponse, error) {
 	if imageID == "" {
 		return image.InspectResponse{}, objectNotFoundError{object: "image", id: imageID}
 	}
diff --git a/client/image_inspect_test.go b/client/image_inspect_test.go
index 0c02dd5d76..9a9f28f0ea 100644
--- a/client/image_inspect_test.go
+++ b/client/image_inspect_test.go
@@ -12,6 +12,7 @@ import (
 
 	cerrdefs "github.com/containerd/errdefs"
 	"github.com/moby/moby/api/types/image"
+	"github.com/moby/moby/client/imageopt"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"gotest.tools/v3/assert"
 	is "gotest.tools/v3/assert/cmp"
@@ -105,7 +106,7 @@ func TestImageInspectWithPlatform(t *testing.T) {
 	}))
 	assert.NilError(t, err)
 
-	var opts []ImageInspectOption
+	var opts []imageopt.InspectOption
 	if requestedPlatform != nil {
 		opts = append(opts, WithPlatform(*requestedPlatform))
 	}
diff --git a/client/imageopt/inspect.go b/client/imageopt/inspect.go
index 668dc55d66..3ad7287a14 100644
--- a/client/imageopt/inspect.go
+++ b/client/imageopt/inspect.go
@@ -5,7 +5,6 @@ import (
 	"context"
 	"fmt"
 
-	"github.com/moby/moby/client"
 	"github.com/moby/moby/client/internal/opts"
 )
 
@@ -20,7 +19,7 @@ func (o *withSinglePlatformOption) ApplyImageInspectOption(ctx context.Context,
 
 // WithRawResponse instructs the client to additionally store the
 // raw inspect response in the provided buffer.
-func WithRawResponse(raw *bytes.Buffer) client.ImageInspectOption {
+func WithRawResponse(raw *bytes.Buffer) InspectOption {
 	return opts.InspectOptionFunc(func(_ context.Context, opts *opts.ImageInspectOptions) error {
 		opts.Raw = raw
 		return nil
@@ -32,7 +31,7 @@ func WithRawResponse(raw *bytes.Buffer) client.ImageInspectOption {
 // With this option set, the image inspect operation response includes
 // the [image.InspectResponse.Manifests] field if the server is multi-platform
 // capable.
-func WithImageManifests(manifests bool) client.ImageInspectOption {
+func WithImageManifests(manifests bool) InspectOption {
 	return opts.InspectOptionFunc(func(_ context.Context, clientOpts *opts.ImageInspectOptions) error {
 		clientOpts.ApiOptions.Manifests = manifests
 		return nil
diff --git a/client/imageopt/opt.go b/client/imageopt/opt.go
new file mode 100644
index 0000000000..33f201ae03
--- /dev/null
+++ b/client/imageopt/opt.go
@@ -0,0 +1,12 @@
+package imageopt
+
+import (
+	"context"
+
+	"github.com/moby/moby/client/internal/opts"
+)
+
+// InspectOption is a type representing functional options for the image inspect operation.
+type InspectOption interface {
+	ApplyImageInspectOption(ctx context.Context, opts *opts.ImageInspectOptions) error
+}
diff --git a/client/imageopt/platform.go b/client/imageopt/platform.go
index 25ffb709c4..483b156a98 100644
--- a/client/imageopt/platform.go
+++ b/client/imageopt/platform.go
@@ -7,11 +7,10 @@ import (
 	"github.com/containerd/platforms"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 
-	"github.com/moby/moby/client"
 	"github.com/moby/moby/client/internal/opts"
 )
 
-func WithPlatformString(platform string) client.ImageInspectOption {
+func WithPlatformString(platform string) InspectOption {
 	return opts.InspectOptionFunc(func(ctx context.Context, opts *opts.ImageInspectOptions) error {
 		p, err := platforms.Parse(platform)
 		if err != nil {
-- 
2.44.0
```

vvoland committed:
Use generic opt

If we won't have shared opts, we can just use a single interface with
generic options struct.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
Revert "Use generic opt"

This reverts commit 2912f917e6fdc765bf58aac4cce0552dc6672341.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
Shared WithPlatformString

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
integration: Adjust ImageHistory and Inspect opts usage

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
client: refactor ContainerStats to functional options

Replaces the boolean `stream` parameter in ContainerStats with a
flexible options-based approach using functional options, and
consolidates `ContainerStats` and `ContainerStatsOneShot`.

The new API introduces:
- `WithStream()` option for streaming stats to a channel
- `WithOneshot()` option for single stat retrieval

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah reviewed (commented):
just some comments; I started to look at this again earlier today, but didn't get far 🙈

vvoland commented on client/container_stats.go:
```diff
@@ -2,66 +2,94 @@ package client
-	query.Set("one-shot", "1")
+		// TODO(vvoland): Should we close the output stream or leave that up to the caller?
+		// It's convenient to close the output stream here so we don't force caller to do it,
+		// but it's a bit inconsistent since the channel was created by the caller.
```
Not sure if iterators are right for this. With channels the user has the ability to affect how the items are buffered (via creating a desired channel) and use `select` to control the consumption.


---

Also, adding it to the returned object would break the uniformity of "sync" vs "async" behavior across the methods (if we went with sync by default) because that would require the method to become "async" (so the user can iterate over the returned object):

https://github.com/moby/moby/blob/9ee0bbd583b7b79faebc98d411bf8b8568b94cfc/client/containerstats/stats.go#L38-L40

->

```go
type Output interface {
	All() iter.Seq2[container.StatsResponse, error]
}
```

So I guess we'd have to do it via functional opt:

```go
func WithIter(out *opts.ContainerStatsIterator) Option {
   ...
}
```

```go
var iter containerstats.Iter
go func() {
	_, err := apiClient.ContainerStats(ctx, cID,
		containerstats.WithIter(&iter),
	)
	assert.NilError(t, err)
}()

for stat, err := range iter.All() {
	assert.NilError(t, err)
	assert.Check(t, stat.MemoryStats.Limit > 0)
	assert.Check(t, !reflect.DeepEqual(stat.PreCPUStats, containertypes.CPUStats{}))
}
```
